File: web/src/types/library.ts
/** 音乐库相关类型，与服务端 schemas 一一对应。 */

export interface Artist {
  id: string;
  name: string;
  album_count: number | null;
  song_count: number | null;
  cover_art: string | null;
}

export interface Album {
  id: string;
  name: string;
  artist_id: string | null;
  artist_name: string | null;
  cover_art: string | null;
  song_count: number | null;
  duration: number | null;
  year: number | null;
  genre: string | null;
}

export interface Song {
  id: string;
  title: string;
  album_id: string | null;
  album_name: string | null;
  artist_id: string | null;
  artist_name: string | null;
  track: number | null;
  year: number | null;
  genre: string | null;
  duration: number | null;
  bit_rate: number | null;
  size: number | null;
  content_type: string | null;
  suffix: string | null;
  path: string | null;
  cover_art: string | null;
}

export interface ArtistDetail extends Artist {
  albums: Album[];
}

export interface AlbumDetail extends Album {
  songs: Song[];
}

export interface SyncState {
  id: number | null;
  scope: string;
  status: string;
  started_at: string;
  finished_at: string | null;
  artists_synced: number;
  albums_synced: number;
  songs_synced: number;
  error: string | null;
}

export interface ArtistList {
  items: Artist[];
  total: number;
  limit: number;
  offset: number;
}

export interface AlbumList {
  items: Album[];
  total: number;
  limit: number;
  offset: number;
}

export interface SongList {
  items: Song[];
  total: number;
  limit: number;
  offset: number;
}

export interface ListQuery {
  q?: string;
  artist_id?: string;
  album_id?: string;
  limit?: number;
  offset?: number;
}

File: server/app/schemas/library.py
"""音乐库相关 API 的响应模型。

与数据库表解耦：前端只看到这里定义的字段，未来表结构变了也不影响契约。
所有模型都 ``from_attributes=True``，可直接从 SQLModel 行实例化。
"""

from __future__ import annotations

from datetime import datetime
from typing import Optional

from pydantic import BaseModel


class ArtistOut(BaseModel):
    model_config = {"from_attributes": True}

    id: str
    name: str
    album_count: Optional[int] = None
    song_count: Optional[int] = None
    cover_art: Optional[str] = None


class AlbumOut(BaseModel):
    model_config = {"from_attributes": True}

    id: str
    name: str
    artist_id: Optional[str] = None
    artist_name: Optional[str] = None
    cover_art: Optional[str] = None
    song_count: Optional[int] = None
    duration: Optional[int] = None
    year: Optional[int] = None
    genre: Optional[str] = None


class SongOut(BaseModel):
    model_config = {"from_attributes": True}

    id: str
    title: str
    album_id: Optional[str] = None
    album_name: Optional[str] = None
    artist_id: Optional[str] = None
    artist_name: Optional[str] = None
    track: Optional[int] = None
    year: Optional[int] = None
    genre: Optional[str] = None
    duration: Optional[int] = None
    bit_rate: Optional[int] = None
    size: Optional[int] = None
    content_type: Optional[str] = None
    suffix: Optional[str] = None
    path: Optional[str] = None
    cover_art: Optional[str] = None


class ArtistDetailOut(ArtistOut):
    albums: list[AlbumOut] = []


class AlbumDetailOut(AlbumOut):
    songs: list[SongOut] = []


class SyncStateOut(BaseModel):
    model_config = {"from_attributes": True}

    id: Optional[int] = None
    scope: str
    status: str
    started_at: datetime
    finished_at: Optional[datetime] = None
    artists_synced: int = 0
    albums_synced: int = 0
    songs_synced: int = 0
    error: Optional[str] = None


class ArtistListOut(BaseModel):
    items: list[ArtistOut]
    total: int
    limit: int
    offset: int


class AlbumListOut(BaseModel):
    items: list[AlbumOut]
    total: int
    limit: int
    offset: int


class SongListOut(BaseModel):
    items: list[SongOut]
    total: int
    limit: int
    offset: int
